hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/CompactingMemStore.java

```diff
@@ -476,55 +506,97 @@ private void doCompact() {
+     * Creates a new Segment with CellFlatMap from the given single old
+     * Segment based on ConcurrentSkipListMap
      */
-    private StoreScanner createScanner(Store store) throws IOException {
-
-      Scan scan = new Scan();
-      scan.setMaxVersions();  //Get all available versions
+    private void flattenSegment(ImmutableSegment segment, int numOfCells) throws IOException {
 
-      StoreScanner internalScanner =
-          new StoreScanner(store, store.getScanInfo(), scan, Collections.singletonList(scanner),
-              ScanType.COMPACT_RETAIN_DELETES, smallestReadPoint, HConstants.OLDEST_TIMESTAMP);
+      LOG.debug("Flattening Segment  " + segment);
+      // Update the pipeline scanner (list of scanners) to scan only Skip List based Segment
+      List<SegmentScanner> scanners = new ArrayList<SegmentScanner>();
+      scanners.add(segment.getSegmentScanner(store.getSmallestReadPoint()));
+      ImmutableSegment newSegment = createFlatFromSegmentWithSkipList(numOfCells, segment);
+      pipeline.replace(versionedList, segment, newSegment);
 
-      return internalScanner;
+      return;
     }
 
     /**
+     * COMPACT TO ARRAY MAP OR TO CHUNK MAP
      * Updates the given single Segment using the internal store scanner,
      * who in turn uses ScanQueryMatcher
      */
-    private void compactSegments(Segment result) throws IOException {
-
-      List<Cell> kvs = new ArrayList<Cell>();
-      // get the limit to the size of the groups to be returned by compactingScanner
-      int compactionKVMax = getConfiguration().getInt(
-          HConstants.COMPACTION_KV_MAX,
-          HConstants.COMPACTION_KV_MAX_DEFAULT);
-
-      ScannerContext scannerContext =
-          ScannerContext.newBuilder().setBatchLimit(compactionKVMax).build();
-
-      boolean hasMore;
-      do {
-        hasMore = compactingScanner.next(kvs, scannerContext);
-        if (!kvs.isEmpty()) {
-          for (Cell c : kvs) {
-            // The scanner is doing all the elimination logic
-            // now we just copy it to the new segment
-            KeyValue kv = KeyValueUtil.ensureKeyValue(c);
-            Cell newKV = result.maybeCloneWithAllocator(kv);
-            result.internalAdd(newKV);
+    private ImmutableSegment compactToFlatSegment(int numOfCells, boolean array)
+        throws IOException {
 
-          }
-          kvs.clear();
-        }
-      } while (hasMore && (!isInterrupted.get()));
+      MemStoreCompactorIterator iterator =
+          new MemStoreCompactorIterator(
+              numOfCells, versionedList.getStoreSegments(), getComparator(), compactionKVMax, store);
+
+      ImmutableSegment result = SegmentFactory.instance()
+          .createImmutableSegment(getConfiguration(), getComparator(), numOfCells, iterator, array);
+
+      iterator.remove();
+      return result;
     }
 
-  }
+    /**
+     * COMPACT TO SKIPLIST MAP
+     */
+    private ImmutableSegment compactToNotFlatSegment(int cells) throws IOException {
```

Comment:
This is actually our current way of compaction. Compacting to ImmutableSegment with CellSet represented as SkipList...